### PR TITLE
1128-api - Limit number of products shown in material cockpit

### PR DESCRIPTION
**Pull request by metas-ts**
* change the product cache for "empty" products to be merged into the cockpit lines from "all" products to ProductFilterVO based LRU caching
* also use the limit sysconfig value

https://github.com/metasfresh/metasfresh-webui-api/issues/1128

### CODE REVIEW

File: src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java [hunk 1]
```diff
@@ -57,10 +63,21 @@
...
-			I_M_Product.Table_Name + "#by#" + I_M_Product.COLUMNNAME_AD_Org_ID,
-			10, // initial size
-			CCache.EXPIREMINUTES_Never);
+	private static final String SYSCONFIG_EMPTY_PRODUCTS_LIMIT = "de.metas.ui.web.material.cockpit.MaterialCockpitRowRepository.EmptyProducts.Limit";
+	private static final int DEFAULT_EMPTY_PRODUCTS_LIMIT = 2000;
+
+	private static final String SYSCONFIG_EMPTY_PRODUCTS_CACHESIZE = "de.metas.ui.web.material.cockpit.MaterialCockpitRowRepository.EmptyProducts.CacheSize";
+	private static final int DEFAULT_EMPTY_PRODUCTS_CACHESIZE = 10;
+
+	private final transient CCache<CacheKey, List<I_M_Product>> productFilterVOToProducts;
```
Comment (teosarca) [inline src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java]: i think a list/set of ProductId(s) would be sufficient.

File: src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java [hunk 2]
```diff
@@ -152,33 +181,50 @@ private boolean retrieveIsIncludePerPlantDetailRows()
...
+				.getOrLoad(cacheKey, () -> retrieveProductsFor(
+						cacheKey.getOrgId(),
+						cacheKey.getProductFilterVO(),
+						cacheKey.getLimit()));
+
+		return products;
 	}
 
-	private List<I_M_Product> retrieveAllProducts(final int adOrgId)
+	private List<I_M_Product> retrieveProductsFor(
```
Comment (teosarca) [inline src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java]: u could have only one param, the CacheKey

Comment (metas-ts) [reply src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java]: i though about that too...yes, doing it..

File: src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java (hunk 1 above)
Comment (metas-ts) [reply src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java]: checked, I think you are right! trying it

File: src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java (hunk 2 above)
Comment (metas-ts) [reply src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java]: done ✔️

File: src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java (hunk 1 above)
Comment (metas-ts) [reply src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java]: done ✔️

**Pull request merged**